sql-plugin/src/main/320+/scala/com/nvidia/spark/rapids/shims/v2/OrcShims.scala

```diff
@@ -0,0 +1,102 @@
+import org.apache.hadoop.hive.common.io.DiskRangeList
+import org.apache.orc.{CompressionCodec, CompressionKind, DataReader, OrcFile, OrcProto, PhysicalWriter, Reader, StripeInformation, TypeDescription}
+import org.apache.orc.impl.{BufferChunk, BufferChunkList, DataReaderProperties, InStream, OrcCodecPool, OutStream, ReaderImpl, SchemaEvolution}
+import org.apache.orc.impl.RecordReaderImpl.SargApplier
+import org.apache.orc.impl.writer.StreamOptions
+
+// 320+ ORC shims
+object OrcShims {
+
+  // read data to buffer
+  def readFileData(dataReader: DataReader, inputDataRanges: DiskRangeList): DiskRangeList = {
+
+    // convert DiskRangeList to BufferChunkList
+    val chuckList = new BufferChunkList
+    var curr = inputDataRanges
+    while (curr != null) {
+      chuckList.add(new BufferChunk(curr.getOffset, curr.getLength))
+      curr = curr.next
+    }
+
+    // BufferChunk is subclass of DiskRangeList
+    dataReader.readFileData(chuckList, false).get().asInstanceOf[DiskRangeList]
+  }
+
+  // create reader properties builder
+  def newDataReaderPropertiesBuilder(compressionSize: Int,
+      compressionKind: CompressionKind, typeCount: Int): DataReaderProperties.Builder = {
+    val compression = new InStream.StreamOptions()
+      .withBufferSize(compressionSize).withCodec(OrcCodecPool.getCodec(compressionKind))
+    DataReaderProperties.builder().withCompression(compression)
+  }
+
+  // create ORC out stream
+  def newOrcOutStream(name: String, bufferSize: Int, codec: CompressionCodec,
+      receiver: PhysicalWriter.OutputReceiver): OutStream = {
+    val options = new StreamOptions(bufferSize).withCodec(codec, codec.getDefaultOptions)
+    new OutStream(name, options, receiver)
+  }
+
+  // filter stripes by pushing down filter
+  def filterStripes(
+      stripes: Seq[StripeInformation],
+      orcReader: Reader,
+      readerOpts: Reader.Options,
+      dataReader: DataReader,
+      updatedReadSchema: TypeDescription,
+      gen: (StripeInformation, OrcProto.StripeFooter, Array[Int], Array[Int]) => OrcOutputStripe,
+      evolution: SchemaEvolution,
+      sargApp: SargApplier,
+      sargColumns: Array[Boolean],
+      ignoreNonUtf8BloomFilter: Boolean,
+      writerVersion: OrcFile.WriterVersion,
+      fileIncluded: Array[Boolean],
+      columnMapping: Array[Int],
+      idMapping: Array[Int]): ArrayBuffer[OrcOutputStripe] = {
+    val result = new ArrayBuffer[OrcOutputStripe](stripes.length)
+
+    // backup the schema
+    val backupSchema = readerOpts.getSchema
+    readerOpts.schema(updatedReadSchema)
+    withResource(new RecordReaderImplSub(orcReader.asInstanceOf[ReaderImpl], readerOpts)) {
+      recordReader =>
+        stripes.foreach { stripe =>
+          val stripeFooter = dataReader.readStripeFooter(stripe)
+
+          recordReader.readRowIndex(stripe.getStripeId.intValue(), fileIncluded, sargColumns)
+          // use returnNone = false underline
+          val rowGroups = recordReader.pickRowGroups()
```

Comment:
`SargApplier` still appears to have the same `pickRowGroups` method in the newer ORC versions, and the old ORC versions have a `RecordReaderImpl.pickRowGroups`, so wondering why this is done differently in Spark 3.2.0+.  Should we be using the record reader approach in previous Spark versions as well, or should Spark 3.2.0+ simply keep using `SargApplier`?  I'm not seeing why there should be a different approach between the two versions.

Note that the predicate pushdown filter is encoded in `sargApp` and otherwise not passed to this method, so curious how predicate pushdown is working with the `RecordReaderImplSub` approach.

Reply:
DataReader.readRowIndex was removed after ORC 1.6.11, so used a different approach.
Actually buildOrcReaderOpts in GpuOrcScanBase created the SearchArgument and then RecordReaderImplSub generated a sargApp.
Now aligned, replaced  DataReader.readRowIndex with StripePlanner.readRowIndex